torch/fx/passes/runtime_assert.py

```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
             # doesn't really have to be in the loop as it will only run once,
             # it just needs to happen right after the placeholders.
             # insert this after placeholders & added sym nodes, and before non-placeholders.
-            if node not in placeholders:
-                last_sym_node = last_placeholder
-                for _ in range(inserted_sym_nodes):
-                    last_sym_node = last_sym_node.next
-                with graph.inserting_before(last_sym_node.next):
-                    add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+            if node == first_non_placeholder:
+                add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+
+            # deduplicate asserts already present in graph
+            if node.target in (
+                torch._check,
+                torch.ops.aten._assert_scalar.default,
+            ):
+                if (
+                    node.args[0] == True  # trivial
+                    or (assert_expr := _get_sym_val(node.args[0])) in added_asserts
+                ):
+                    gm.graph.erase_node(node)
+                else:
+                    added_asserts.add(assert_expr)
+
+            # hash cons on symbolic values
+            if (
+                node.op == "call_function"
+                and (sym_expr := _get_sym_val(node)) is not None
+            ):
+                if (
+                    sym_expr in symbol_to_proxy  # example value is redundant
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
+                    )  # shape call on intermediate tensor, turn into computation on input shapes
```

Comment:
Yeah, so concretely, I would just suggest unconditionally generating the sympy expression (hitting memo if necessary) when you get here, instead of gating on the logic here